### DISCUSSION TITLE
v5.0.0RC2

### DISCUSSION THREAD

**Discussion by niden** [phalcon/cphalcon#15988] [General]
v5.0.0RC2

<hr /><em>This discussion was created from the release <a href='https://github.com/phalcon/cphalcon/releases/tag/v5.0.0RC2'>v5.0.0RC2</a>.</em>

**Comment by s-ohnishi** [comment]
Since v5 is the RC version, I'd like to try it, but the online manual is still v4, and the phalcon/mvc sample can not be used as it is due to a series of errors.
Is there a sample that works with v5?
Since it is RC, I do not want detailed explanations.
It takes a lot of time and effort to search for changes in namespaces, method names, and arguments by relying on .zep and stub, so I need something that can be used as a reference.

Or should just non-developer users wait for the manual to be available?

**Reply by Jeckerson** [reply]
You can try it by using changelog - https://github.com/phalcon/cphalcon/blob/master/CHANGELOG-5.0.md

**Reply by s-ohnishi** [reply]
Really?
Do I have to look it up one by one?

**Reply by Jeckerson** [reply]
> Really? Do I have to look it up one by one?

In case you are in rush, while docs are in progress...

**Reply by s-ohnishi** [reply]
Even if it wasn't completed as a document, I wish I had a sample that worked fine ...

**Reply by niden** [reply]
We are trying our best to get the documents done. As you know we are a small team so it takes a lot longer than other frameworks that have a lot more resources. Sorry about that but we are doing all we can to get this version out as fast as possible.

**Reply by s-ohnishi** [reply]
I am aware of these circumstances.
It seems that you guys think "I want you to finish the document quickly", but I don't mean to do that.
There is no need for explanations, so if there are "working samples", I can improve their understanding by examining them.
It seems that there are many changes, and it takes a lot of time to port existing samples (various patterns and tutorials of phalcon/mvc and tutorials) while comparing CHANGELOG and .zep.
If there is no sample that can be presented, I will give up and wait.

I'm not trying to force you to work.
It takes time to understand v5, so I just wanted to practice with the RC version.

**Reply by Jeckerson** [reply]
@s-ohnishi yes

**Reply by s-ohnishi** [reply]
OK.
I will go back to v4.

**Reply by kgrammer** [reply]
I used the CHANGELOG to go through what I needed to change.  It took about an hour, but I will admit that for my site, I am using a smaller sub-set of Phalcon then some.  I do manage an LMS that has more Phalcon calls, so I would estimate that codebase to require perhaps two hours to review.

I posted about what I had to change over in the General channel on Discord.

**Reply by kgrammer** [reply]
Niden,

After reading this discussion, I pulled the tutorial files down and made a few updates and got them working with just fine with Phalcon 5.0.0RC2.

I struggled a bit with the Phalcon\Tags::linkTo() for two reasons.  First, based on something I thought I read in the CHANGELOG, I *thought* Phalcon\Tags was replaced with Phalcon\Html\TagFactory.  Second, I'm not a phtml guy.  I've always used VOLT files so I struggled for a few minutes to find the proper phtml syntax.  

This raised the question of should the tutorial stick with phtml files, or should it be updated to use VOLT files?

Anyway, I can post or email you a ZIP file if someone would like to review the tutorial files.

**Reply by niden** [reply]
@kgrammer Yes we are using the `TagFactory` from now on (or should be anyway), but the `Tag` component was left there because if we take it out it will pretty much destroy applications and force people to do full rewrites. Leaving it there and giving developers a full version to adjust their applications is the best way forward.

The `linkTo` can be replaced by the `Anchor` class.

Volt:

```twig
{{ tag.a('/myurl', 'click me', ['class' : 'label']) }}
```
and in phtml it will become

```php
<?php echo $this->tag->a('/myurl', 'click me', ['class' => 'label']); ?>
```

Yes please for the tutorial. Let's communicate and we can get that updated for v5. I am in parallel working to get Vokuro done with 5.x

**Reply by kgrammer** [reply]
Cool.  Thanks.

What would be the correct code to for the form?

I have the tutorial working with the following:

```
Phalcon\Tag::form('signup/register')
Phalcon\Tag::textField("name")
Phalcon\Tag::textField("email")
Phalcon\Tag::submitButton(["Register", 'class' => 'btn btn-primary'])
Phalcon\Tag::endForm()
```
I use straight VOLT/HTML in my sites so I'm unfamiliar with the Phalcon Tag manager syntax.

Thanks.

**Reply by niden** [reply]
This should be

```php
$this->tag->form(['action' => '/signup/register']);
$this->tag->inputText('name');
$this->tag->inputText('email');
$this->tag->inputSubmit('Register, null, ['class' => 'btn btn-primary']);
$this->tag->close('form');
```

for Volt 

```twig
{{ tag.form(['action' : '/signup/register']) }}
{{ tag.inputText('name') }}
{{ tag.inputText('email') }}
{{ tag.inputSubmit('Register', null, ['class' : 'btn btn-primary']) }}
{{ tag.close('form') }}
```

**Reply by kgrammer** [reply]
Perfect.  That fixed it... (once I added the missing quote after "register" on the inputSubmit() tag).  :D

Thanks.

**Reply by s-ohnishi** [reply]
> for Volt
> 
> ```
> {{ tag.form(['action' : '/signup/register']) }}
> {{ tag.inputText('name') }}
> {{ tag.inputText('email') }}
> {{ tag.inputSubmit('Register, null, ['class' : 'btn btn-primary']) }}
> {{ tag.close('form') }}
> ```

This is the case when you write it directly in .volt, right?
What if I want to specify a stylesheet or JavaScript in the controller?


for example.
in the Controller
```
$this->tag->style()->add('css/some.css');
```
in the Volt
```
{{ tag.style() }}
```
This will not show anything.

in the Volt
```
<?php var_dump($this->tag); ?>
```
It is displayed like this (partially excerpt)
```
'style' => 
        object(Phalcon\Html\Helper\Style)[100]
          protected 'delimiter' => string '
' (length=1)
          protected 'escaper' => 
            object(Phalcon\Html\Escaper)[102]
              protected 'doubleEncode' => boolean true
              protected 'encoding' => string 'utf-8' (length=5)
              protected 'flags' => int 11
          protected 'indent' => string '    ' (length=4)
          protected 'indentLevel' => int 1
          protected 'attributes' => 
            array (size=0)
              empty
          protected 'store' => 
            array (size=1)
              0 => 
                array (size=3)
                  0 => string 'renderTag' (length=9)
                  1 => 
                    array (size=3)
                      0 => string 'link' (length=4)
                      1 => 
                        array (size=4)
                          'rel' => string 'stylesheet' (length=10)
                          'href' => string 'css/some.css' (length=12)
                          'type' => string 'text/css' (length=8)
                          'media' => string 'screen' (length=6)
                      2 => string '/' (length=1)
                  2 => string '    ' (length=4)
          private 'isStyle' => boolean false
```

**Reply by s-ohnishi** [reply]
It seems that stylesheet and JavaScript can be registered multiple record (multiple registrations are made to the sotre property), but isn't it only the last value that is retained even if the method shown above is repeated?
How do I register more than one?

**Reply by s-ohnishi** [reply]
I'm sorry, the multiple registrations have been resolved.
in the Controller
```
$style = $this->tag->style();
$style->add('some.css);
$style->add('other.css);
```
However, the output with Volt remains unclear.

And, do I register after going through the url when registering?
Or will it automatically pass the url at the time of output?


The title works well below.
in the Controller
``` 
$this->tag->title()->prepend(SITE_NAME)->setSeparator(' - ');
```
in the Volt
```
{{ tag.title() }}
```

Besides, I can't even register with doctype, and I can't retrieve it with script as well as style.

**Reply by s-ohnishi** [reply]
I found a way to take it out.
```
{{tag.style ()}}
{{tag.script ()}}
```
Nothing was displayed, but you can retrieve it by writing as follows.
```{{tag.newInstance ('style')}}
{{tag.newInstance ('script')}}
```
It feels strange that it is "new" even though it has already been registered.

And I thought that TagFactory::__call() would return the same result because it can be read as if it were just calling newInstance() internally, but it was different as above.

By the way, {{ tag.title() }} works.
but...
```
@method title (string $ separator ='', string $ indent ='', string $ delimiter = PHP_EOL): string
```
[TagFactory.zep](https://github.com/phalcon/cphalcon/blob/5.0.x/phalcon/Html/TagFactory.zep)
Therefore, if you specify three arguments, an error will occur.
Fatal error: Uncaught ArgumentCountError: Phalcon\Html\Helper\Title::__invoke() expects at most 2 arguments, 3 given

And even if you pass a parameter like
```
tag.title("\t", "\n\n")
```
 it will be escaped, so it will not work as expected.

**Reply by niden** [reply]
@s-ohnishi Thank you for reporting this. 

For the `title()` in the TagFactory, that is a typo which I have created. The correct signature is:

```
@method title (string $ indent = '    ', string $ delimiter = PHP_EOL): string
```

The `newInstance()` in the `TagFactory` creates a new object if it does not exist or returns the existing object during the same request.

I will create some tests to check why script etc. do not output anything and also the title one.

**Reply by s-ohnishi** [reply]
I am very much looking forward to improvement.

Also, even if I specify $indent and $delimiter, .volt is replaced with single quotes at compile time, so why can't I pass tabs or line breaks?

in the Volt
```
{{ tag.title("\t\t","\n\n\n\n") }}
```
compile to cache
```
<?= $this->tag->title('\t\t', '\n\n\n\n') ?>
```

**Reply by niden** [reply]
@s-ohnishi Have a look if I missed anything please: https://github.com/phalcon/cphalcon/issues/16005

**Reply by s-ohnishi** [reply]
That's about right.
I'm just starting to port, so I haven't tried any other HTML elements that I control in my controller (such as meta).

**Reply by niden** [reply]
@s-ohnishi I made necessary corrections and wrote tests to see that the code from `style()` etc. is correctly generated using Volt. This will be out in the next RC which will be this week.

Thank you for reporting this.

**Reply by s-ohnishi** [reply]
That's good news.
I am very much looking forward.

**Reply by s-ohnishi** [reply]
I tried RC3.
I was able to output the value set in the controller in .volt.
Thank you very much.

```
SomeController::
$this->tag->doctype(\Phalcon\Tag::HTML401_TRANSITIONAL, PHP_EOL);

.volt::
{{ tag.newInstance('doctype') }}
```

But when I use {{ tag.doctype() }} in .volt, <! DOCTYPE html> is returned.

I may not understand Zephir exactly
```
TagFactory::
    public function __call(string name, array arguments)
    {
        var helper;
        let helper = this->newInstance(name);
        return call_user_func_array([helper, "__invoke"], arguments);
    }

Doctype::
    public function __invoke(
        int flag = self::HTML5,
        string delimiter = "\n"
    ) -> <Doctype> {
        let this->flag      = flag,
        this->delimiter = delimiter;
        return this;
    }
```

Doesn't this always return a new instance with tag.doctype()?

For example, if you make it like this, tag.tagname () will return the instance if it is registered, register the instance using arguments if it is not registered, and return it?
And in order to make newInstance () a feature that matches the name, I think we should just overwrite the existing service with a "new" instance and return it.

```
TagFactory::

    public function __call(string name, array arguments = [])
    {
        var helper;
        if !isset this->services[name] {
            let helper = this->newInstance(name);
            this->services[name] = call_user_func_array([helper, "__invoke"], arguments);
        }
        return this->services[name];
    }

    public function newInstance(string name) -> var
    {
        var definition;

        let definition = this->getService(name),
        this->services[name] = create_instance_params(
            definition,
            [
                this->escaper
            ]
        );

        return this->services[name];
    }
```

I don't understand zephir, and I don't know how it affects others, but I hope it conveys what I mean.

**Reply by niden** [reply]
This seems indeed a bit weird. Let me test it a bit and explain it to my head first and then I will share lol.

In theory, the TagFactory is a locator service that creates objects and stores them to be reused within the same request. The `newInstance` creates a new object only if not present - that is the design at least.

There is something I don't like in the `__call` method. I need to check it.

Also for the `doctype` method you use, it should return the transitional string. Let me test it first and will post results shortly.

**Reply by s-ohnishi** [reply]
In the current situation, the __call method always returns a new instance, and newInstance() may return an existing (= old) instance, so I think the function is the opposite.
And current newInstance() can only pass the name, it feels inconvenient not to pass a parameter.
(I forgot to include that part in the previous reply)

I will wait for your test results to be shown.

**Comment by kgrammer** [comment]
@niden, I have the tutorial file for Phalcon 5 ready to send to you.  You can message me on Discord with instructions on where you would like me to send the zip file.

**Comment by lieshanhe** [comment]
FYI, if you can not find the documentation and occurred some Fatal Error like 

`Fatal Error: Class "Phalcon\Config" not found`

you can go to the source page
https://github.com/phalcon/cphalcon/tree/master/phalcon

and find the path of the namespace, it's straightforward.

**Reply by s-ohnishi** [reply]
Thank you very much.
I know that page, and I often follow it from that page when I want to find out what parameters I can specify and what I'm doing.
However, in this case, the problem was that "nothing was displayed" instead of an error.


PhpStorm does a good job, so you don't have to worry too much about namespaces.
But, sometimes the stub is wrong and displays a warning.
I found another one.
```
  * @method label (array $ attributes = []): string
```
phalcon / Html / Helper / Label.zep
```
public function __invoke (string $ label, array $ attributes = [], bool $ raw = false): string
```
